index.html

```diff
@@ -1,49 +1,256 @@
+        {"state":"MT","rate":17.2},
+        {"state":"NC","rate":11.7},
+        {"state":"ND","rate":17.4},
+        {"state":"NE","rate":16.2},
+        {"state":"NH","rate":10.4},
+        {"state":"NJ","rate":13.1},
+        {"state":"NM","rate":13.2},
+        {"state":"NV","rate":37.8},
+        {"state":"NY","rate":41.0},
+        {"state":"OH","rate":10.2},
+        {"state":"OK","rate":10.9},
+        {"state":"OR","rate":31.0},
+        {"state":"PA","rate":12.0},
+        {"state":"PR","rate":11.4},
+        {"state":"RI","rate":11.3},
+        {"state":"SC","rate":10.6},
+        {"state":"SD","rate":10.5},
+        {"state":"TN","rate":14.5},
+        {"state":"TX","rate":10.8},
+        {"state":"UT","rate":10.6},
+        {"state":"VA","rate":8.5},
+        {"state":"VT","rate":24.9},
+        {"state":"WA","rate":26.5},
+        {"state":"WI","rate":10.5},
+        {"state":"WV","rate":10.9},
+        {"state":"WY","rate":13.0}
+      ]; 
+
+      var MAX_RATE = 49.3; 
+
+      var options = {
+      	sorted: false, 
+      	filtered: false, 
+      	page: 0
+      }; 
+
+      var tbody = d3.select("tbody"); 
+
+      function updateTable() {
+
+        tbody.html(""); 
+
+        var data = states.slice()
+
+        if (options.filtered === 'filter-r') { 
+            data = data.filter(function (a) { 
+            return a.rate > 15}); 
+        }
+
+        var rows = tbody.selectAll("tr")
+          .data(states)
+          .enter()
+          .append("tr"); 
+
+        rows.append("td")
+          .html(function(d) {return d.state}); 
+        
+        rows.append("td")
+          .html(function (d) {return d.rate}); 
+
+        rows.append("td")
+          .append("div")
+          .classed("bar", true)
+          .style("width", function (d) {return ((d.rate / MAX_RATE) *100) + '%'})
+
+        }; 
+
+        updateTable(); 
+
+        d3.select('#sort-asc')
+          .on("click", function() {
+
+            if (d3.select('#sort-asc').classed('active') === false) {
+              d3.select('#sort-asc').classed('active', true);
+              d3.select('#sort-desc').classed('active', false); 
+              d3.selectAll("tbody tr").sort(function (a, b) {
+                return d3.ascending(a.rate, b.rate); 
+              }); 
+            } else { 
+              d3.select('#sort-asc').classed('active', false); 
+
+            updateTable(); 
+            }
+
+          }); 
+
+        d3.select('#sort-desc')
+          .on("click", function() {
+
+            if (d3.select('#sort-desc').classed('active') === false) {
+              d3.select('#sort-desc').classed('active', true);
+              d3.select('#sort-asc').classed('active', false); 
+              d3.selectAll("tbody tr").sort(function (a, b) {
+                return d3.descending(a.rate, b.rate); 
+              }); 
+            } else { 
+              d3.select('#sort-desc').classed('active', false); 
+
+            updateTable(); 
+            }
```

Comment:
So the above code works for sorting, though it wasn't the way we originally imagined.

We were trying to get you to grab the code from last week (week 2) that transforms the data based on options within the updateTable() function. So the sort, in addition to the filter, would have been inside the updateTable() function. The buttons themselves would then simply change the options and rerun updateTable(). That's what Chris was trying to get to in his email on Wednesday.

The reason this is important is that sort and filter work differently. Sort is going to rearrange the DOM elements all on its own. Filter isn't going to do this however, it just filters the selection that the data is bound to. This is why we need to filter, then run updateTable again - clearing and repopulating all the table elements.